contrib/endpoints/src/api_manager/request_handler_test.cc

```diff
@@ -0,0 +1,619 @@
+  int64_t GetGrpcRequestBytes() { return 0; }
+  int64_t GetGrpcResponseBytes() { return 0; }
+  int64_t GetGrpcRequestMessageCounts() { return 0; }
+  int64_t GetGrpcResponseMessageCounts() { return 0; }
+
+  bool FindQuery(const std::string &name, std::string *query) {
+    if (data_.find("query." + name) == data_.end()) {
+      return false;
+    }
+    *query = data_["query." + name];
+    return true;
+  }
+
+  bool FindHeader(const std::string &name, std::string *header) {
+    if (data_.find("header." + name) == data_.end()) {
+      return false;
+    }
+    *header = data_["header." + name];
+    return true;
+  }
+
+  ::google::api_manager::protocol::Protocol GetFrontendProtocol() {
+    return ::google::api_manager::protocol::Protocol::HTTP;
+  }
+
+  ::google::api_manager::protocol::Protocol GetBackendProtocol() {
+    return ::google::api_manager::protocol::Protocol::HTTPS;
+  }
+
+  void SetAuthToken(const std::string &auth_token) {}
+
+  utils::Status AddHeaderToBackend(const std::string &key,
+                                   const std::string &value) {
+    return utils::Status::OK;
+  }
+
+ private:
+  std::unordered_map<std::string, std::string> data_;
+};
+
+class ResponseMock : public Response {
+ public:
+  virtual ~ResponseMock() {}
+  utils::Status GetResponseStatus() { return utils::Status::OK; }
+  std::size_t GetRequestSize() { return 0; }
+  std::size_t GetResponseSize() { return 0; }
+  utils::Status GetLatencyInfo(service_control::LatencyInfo *info) {
+    return utils::Status::OK;
+  }
+};
+
+class RequestHandlerTest : public ::testing::Test {
+ protected:
+  RequestHandlerTest() : callback_run_count_(0) {}
+  std::shared_ptr<ApiManager> MakeApiManager(
+      std::unique_ptr<ApiManagerEnvInterface> env, const char *service_config);
+  std::shared_ptr<ApiManager> MakeApiManager(
+      std::unique_ptr<ApiManagerEnvInterface> env, const char *service_config,
+      const char *server_config);
+
+  void SetUp() {
+    callback_run_count_ = 0;
+    call_history_.clear();
+  }
+
+ protected:
+  std::vector<std::string> call_history_;
+  int callback_run_count_;
+
+ private:
+  ApiManagerFactory factory_;
+};
+
+std::shared_ptr<ApiManager> RequestHandlerTest::MakeApiManager(
+    std::unique_ptr<ApiManagerEnvInterface> env, const char *service_config) {
+  return factory_.CreateApiManager(std::move(env), service_config, "");
+}
+
+std::shared_ptr<ApiManager> RequestHandlerTest::MakeApiManager(
+    std::unique_ptr<ApiManagerEnvInterface> env, const char *service_config,
+    const char *server_config) {
+  return factory_.CreateApiManager(std::move(env), service_config,
+                                   server_config);
+}
+
+TEST_F(RequestHandlerTest, PendingCheckApiManagerInitSucceeded) {
+  std::unique_ptr<MockApiManagerEnvironment> env(
+      new ::testing::NiceMock<MockApiManagerEnvironment>());
+
+  EXPECT_CALL(*(env.get()), DoRunHTTPRequest(_))
+      .WillRepeatedly(Invoke([this](HTTPRequest *req) {
```

Comment:
Since you know the sequence of calls, so it is better to call in this form:

EXPECT_CALL().
     .Invoke(  you can assert specific url and return its boby )
     .Invoke( the second call).